File: apollo-runtime/src/test/java/com/apollographql/apollo/internal/ResponseFetcherTest.java
package com.apollographql.apollo.internal;

import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.OperationName;
import com.apollographql.apollo.api.Query;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.api.ResponseReader;
import com.apollographql.apollo.api.cache.http.HttpCachePolicy;

import org.junit.Before;
import org.junit.Test;

import org.jetbrains.annotations.NotNull;

import okhttp3.OkHttpClient;

import static com.apollographql.apollo.fetcher.ApolloResponseFetchers.CACHE_FIRST;
import static com.apollographql.apollo.fetcher.ApolloResponseFetchers.NETWORK_ONLY;
import static com.google.common.truth.Truth.assertThat;

public class ResponseFetcherTest {
  private OkHttpClient okHttpClient;
  private Query emptyQuery;

  @Before public void setUp() {
    okHttpClient = new OkHttpClient.Builder().build();

    emptyQuery = new Query() {
      OperationName operationName = new OperationName() {
        @Override public String name() {
          return "emptyQuery";
        }
      };
      @Override public String queryDocument() {
        return "";
      }

      @Override public Variables variables() {
        return EMPTY_VARIABLES;
      }

      @Override public ResponseFieldMapper<Data> responseFieldMapper() {
        return new ResponseFieldMapper<Data>() {
          @Override public Data map(ResponseReader responseReader) {
            return null;
          }
        };
      }

      @NotNull @Override public OperationName name() {
        return operationName;
      }

      @NotNull @Override public String operationId() {
        return "";
      }

      @Override public Object wrapData(Data data) {
        return data;
      }
    };
  }

  @Test public void setDefaultCachePolicy() {
    ApolloClient apolloClient = ApolloClient.builder()
        .serverUrl("http://google.com")
        .okHttpClient(okHttpClient)
        .defaultHttpCachePolicy(HttpCachePolicy.CACHE_ONLY)
        .defaultResponseFetcher(NETWORK_ONLY)
        .build();

    RealApolloCall realApolloCall = (RealApolloCall) apolloClient.query(emptyQuery);
    assertThat(realApolloCall.httpCachePolicy.fetchStrategy).isEqualTo(HttpCachePolicy.FetchStrategy.CACHE_ONLY);
    assertThat(realApolloCall.responseFetcher).isEqualTo(NETWORK_ONLY);
  }

  @Test public void defaultCacheControl() {
    ApolloClient apolloClient = ApolloClient.builder()
        .serverUrl("http://google.com")
        .okHttpClient(okHttpClient)
        .build();

    RealApolloCall realApolloCall = (RealApolloCall) apolloClient.query(emptyQuery);
    assertThat(realApolloCall.httpCachePolicy.fetchStrategy).isEqualTo(HttpCachePolicy.FetchStrategy.NETWORK_ONLY);
    assertThat(realApolloCall.responseFetcher).isEqualTo(CACHE_FIRST);
  }
}

File: apollo-runtime/src/test/java/com/apollographql/apollo/internal/ResponseFetcherTest.kt
package com.apollographql.apollo.internal

import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.Query
import com.apollographql.apollo.api.cache.http.HttpCachePolicy
import com.apollographql.apollo.api.cache.http.HttpCachePolicy.FetchStrategy
import com.apollographql.apollo.fetcher.ApolloResponseFetchers
import com.google.common.truth.Truth
import okhttp3.OkHttpClient
import org.junit.Test

class ResponseFetcherTest {

  private val emptyQuery = object : Query<Operation.Data> {
    var operationName: String  ="emptyQuery"

    override fun queryDocument(): String {
      return ""
    }

    override fun variables(): Operation.Variables {
      return Operation.EMPTY_VARIABLES
    }

    override fun adapter() = throw UnsupportedOperationException()

    override fun name(): String {
      return operationName
    }

    override fun operationId(): String {
      return ""
    }
  }

  @Test
  fun setDefaultCachePolicy() {
    val apolloClient = ApolloClient.builder()
        .serverUrl("http://google.com")
        .okHttpClient(OkHttpClient())
        .defaultHttpCachePolicy(HttpCachePolicy.CACHE_ONLY)
        .defaultResponseFetcher(ApolloResponseFetchers.NETWORK_ONLY)
        .build()
    val realApolloCall = apolloClient.query(emptyQuery) as RealApolloCall<*>
    Truth.assertThat(realApolloCall.httpCachePolicy!!.fetchStrategy).isEqualTo(FetchStrategy.CACHE_ONLY)
    Truth.assertThat(realApolloCall.responseFetcher).isEqualTo(ApolloResponseFetchers.NETWORK_ONLY)
  }

  @Test
  fun defaultCacheControl() {
    val apolloClient = ApolloClient.builder()
        .serverUrl("http://google.com")
        .okHttpClient(OkHttpClient())
        .build()
    val realApolloCall = apolloClient.query<Operation.Data>(emptyQuery) as RealApolloCall<*>
    Truth.assertThat(realApolloCall.httpCachePolicy!!.fetchStrategy).isEqualTo(FetchStrategy.NETWORK_ONLY)
    Truth.assertThat(realApolloCall.responseFetcher).isEqualTo(ApolloResponseFetchers.CACHE_FIRST)
  }
}
